README
NAME
    Finance::TW::TAIFEX - Helper functions for Taiwan Futures Exchange

SYNOPSIS
      use Finance::TW::TAIFEX;

      my $taifex = Finance::TW::TAIFEX->new();

      $taifex->is_trading_day(); # is today a trading day?

      my $date = DateTime->now;
      $taifex->daily_futures_uri($date);
      $taifex->daily_options_uri($date);

      $taifex->contract('TX', '201001')->settlement_date;
      $taifex->product('TX')->near_term;
      $taifex->product('TX')->next_term;

DESCRIPTION
    Finance::TW::TAIFEX provides useful helper functions for the Taiwan
    Future Exchanges.

METHODS
  product NAME
    Returns the Finance::TW::TAIFEX::Product object represented by NAME.

    Currently supported product names:

    Futures
        TX MTX TE TF T5F MSF CPF XIF GTF

    Options
        TXO TEO TFO MSO XIO GTO

  has_product NAME
    Checks if the given product exists.

  contract NAME YEAR MONTH
    Returns the Finance::TW::TAIFEX::Contract of the given product expires
    on YEAR/MONTH.

  calendar_for YEAR
    Returns the trading calendar for YEAR.

  is_trading_day [DATE]
    Checks if the given DATE is a known trading day. Default DATE is the
    date in the current context.

  next_trading_day [DATE]
    Returns the next known trading day in string after the given DATE. This
    doesn't work across years yet.

  previous_trading_day [DATE]
    Returns the previous known trading day in string after the given DATE.
    This doesn't work across years yet.

  daily_futures_uri DATE
    Returns the URI of the official TAIFEX futures trading records for DATE.

  interday_futures_request($product, [$DATE])
    Returns a HTTP::Request object that fetches futures monthly interday csv
    file for $product of $DATE.

  interday_options_request($product, [$DATE])
    Returns a HTTP::Request object that fetches options monthly interday csv
    file for $product of $DATE.

  daily_options_uri DATE
    Returns the URI of the official TAIFEX options trading records for DATE.

CAVEATS
    The URI returned by "daily_futures_uri" and "daily_options_uri" are only
    valid for the last 30 trading days per the policy of TAIFEX.

AUTHOR
    Chia-liang Kao <clkao@clkao.org>

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    <http://www.taifex.com.tw/>



Changes
Revision history for Perl extension Finance::TW::TAIFEX

0.40 Fri Feb 22 23:18:57 CST 2013
        - Add 2013-02-23 to calendar.

0.39 Sun Dec 30 15:53:15 CST 2012
        - 2013 calendar.

0.38 Fri Aug  3 11:53:01 CST 2012
        - update calender for 2012 typhoon.

0.37 Thu Feb  2 14:51:28 CST 2012
        - fix two missing Saturdays in 2012 calendar.

0.36 Wed Jan 18 19:02:51 CST 2012
        - helper scripts for eod data maintenance.

0.35 Tue Dec 27 18:20:00 CST 2011
        - calendar for 2012.
        - tx helper scripts.
        - make next/previous trading day api work across years.

0.34 Sun May  1 17:33:59 CST 2011
        - correct 2011 calendar.

0.33 Sun Feb  6 21:25:37 CST 2011
        - Better any-moose deps handling.
        - correct 2011 lunar new year calendar.

0.32 Sat Dec 25 20:10:39 CST 2010
        - Use Any::Moose.
        - Calendar for 2011.
        - Calendar for 2000-2008.

0.31 Mon Oct 11 09:28:46 CST 2010
        - Use exchange timezone by default.
        - Fix dependencies.

0.30 Sat Oct  9 17:03:03 CST 2010
        - Use moose native attribute traits.
        - Calendar updates.

0.29  Sat Jan 30 12:03:52 CST 2010
        - Correct calendar for Feb 2010.

0.28  Fri Jan  1 17:47:05 2010
        - original version


Makefile.PL
sub requires_any_moose {};sub readme_from {}; sub auto_include_deps {}; sub author_tests {}; sub auto_set_repository {};
use inc::Module::Install;
readme_from 'lib/Finance/TW/TAIFEX.pm';
name 'Finance-TW-TAIFEX';
all_from 'lib/Finance/TW/TAIFEX.pm';
readme_from 'lib/Finance/TW/TAIFEX.pm';
build_requires 'Test::More';
build_requires 'Test::Exception';

requires 'DateTime';
requires 'DateTime::Format::Strptime';
requires 'Archive::Zip';
requires 'File::ShareDir';
requires 'List::MoreUtils';
requires 'HTTP::Request::Common';
requires 'LWP::Simple';

eval 'require Moose; 1';
if ($@ || -d 'inc/.author') {
    requires 'MouseX::NativeTraits'; # XXX: there's nativetraits are builtin for moose

    requires_any_moose 'X::Traits';
    requires_any_moose 'X::Types::DateTime';
}
else {
    requires 'Any::Moose';
    requires 'MooseX::Traits';
    requires 'MooseX::Types::DateTime';
}

install_share 'share';
install_script glob('bin/*');

auto_include_deps;
author_tests('xt');
auto_set_repository;
WriteAll;


lib/Finance/TW/TAIFEX/Product.pm
package Finance::TW::TAIFEX::Product;
use Any::Moose;

=head1 NAME

Finance::TW::TAIFEX::Contract - Product on TAIFEX

=head1 METHODS

=cut

with any_moose('X::Traits');

has name => (is => "ro", isa => "Str");

has exchange => (is => "ro", isa => "Finance::TW::TAIFEX");

has '+_trait_namespace' => ( default => 'Finance::TW::TAIFEX' );

sub _near_term {
    my ($self, $date) = @_;
    my $settlement = $self->find_settlement_day($date);
    if ( $date->day > $settlement->day ) {
        $settlement = $settlement->add( months => 1);
    }
    return $settlement;
}

=head2 near_term DATE

Returns the near term contract of the product at DATE in YYYYMM format.

=cut

sub near_term {
    my ($self, $date) = @_;
    return $self->_near_term($date)->strftime('%Y%m');
}

=head2 near_term_contract DATE

Returns the near term contract object for product at DATE.

=cut

sub near_term_contract {
    my ($self, $date) = @_;
    my $near_term = $self->_near_term($date);
    return Finance::TW::TAIFEX::Contract->new(
        product => $self,
        year => $near_term->year,
        month => $near_term->month);
}

=head2 find_settlement_day DATE

Find the settlement day of the near term contract at DATE.

=cut

sub find_settlement_day {
    my ($self, $date) = @_;
    my $settlement = $self->default_settlement_day($date);

    while (!$self->exchange->is_trading_day($settlement)) {
        $settlement->add( days => 1 );
    }

    return $settlement
}

__PACKAGE__->meta->make_immutable;
no Any::Moose;
1;


lib/Finance/TW/TAIFEX/Contract.pm
package Finance::TW::TAIFEX::Contract;
use Any::Moose;

=head1 NAME

Finance::TW::TAIFEX::Contract - Contract traded on TAIFEX

=head1 METHODS

=cut

has year  => (is => "ro", isa => "Int", required => 1);
has month => (is => "ro", isa => "Int", required => 1);
has product => (is => "ro", isa => "Finance::TW::TAIFEX::Product",
                handles => [qw(name exchange find_settlement_day)],
            );

=head2 is_settlement_day DATE

Check if DATE is a settlement day

=cut

sub is_settlement_day {
    my ($self, $date) = @_;
    return $date->ymd eq $self->settlement_day->ymd
}

=head2 settlement_day

Returns the settlement day of the contract.

=cut

sub settlement_day {
    my $self = shift;

    $self->find_settlement_day(
        DateTime->new( year => $self->year, month => $self->month, day => 1)
    );
}


my $local_month_code = ['A'..'L'];
sub encode_local {
    my $self = shift;
    return $local_month_code->[$self->month-1].substr($self->year, -1, 1);
}

my $month_code = [qw(F G H J K M N Q U V X Z)];
sub encode {
    my $self = shift;
    return $month_code->[$self->month-1].substr($self->year, -1, 1);
}

__PACKAGE__->meta->make_immutable;
no Any::Moose;
1;



lib/Finance/TW/TAIFEX.pm
package Finance::TW::TAIFEX;
use strict;
use Any::Moose;
use DateTime;
use DateTime::Format::Strptime;
use Try::Tiny;
use File::ShareDir qw(dist_dir);
use List::MoreUtils qw(firstidx);
use Any::Moose 'X::Types::DateTime';
require MouseX::NativeTraits if Any::Moose->mouse_is_preferred;
use HTTP::Request::Common qw(POST);
use LWP::Simple 'getstore';
use Archive::Zip qw( :ERROR_CODES :CONSTANTS );

use Finance::TW::TAIFEX::Product;
use Finance::TW::TAIFEX::Contract;

use 5.008_001;
our $VERSION = '0.40';

has context_date => ( is => "rw", isa => "DateTime",
                      default => sub { DateTime->now(time_zone => 'Asia/Taipei') },
                      coerce => 1);

has calendar => (is => "ro", isa => "HashRef", default => sub { {} });

has products => (
    traits => ['Hash'],
    is => "ro",
    isa => "HashRef[Finance::TW::TAIFEX::Product]",
    handles  => {
        has_product => 'exists',
        product     => 'get',
    },
    lazy_build => 1,
);


sub _build_products {
    my $self = shift;
    return {
        (map { $_ => Finance::TW::TAIFEX::Product->new_with_traits(
            traits => ['Settlement::ThirdWednesday'],
            exchange => $self,
            name => $_,
        ) } qw(TX MTX TE TF T5F MSF CPF XIF GTF
               TXO TEO TFO MSO XIO GTO)),
    };

#        (map { $_ => Finance::TW::TAIFEX::Product->new_with_traits(
#            traits => ['Settlement::ThirdToLastDayOfMonth'],
#            exchange => $self,
#            name => $_,
#        ) } qw(TGF TGO)),
#
#        (GBF => Finance::TW::TAIFEX::Product->new_with_traits(
#            traits => ['Settlement::SecondWednesday'],
#            exchange => $self,
#            name => 'GBF',
#        )),

}

=head1 NAME

Finance::TW::TAIFEX - Helper functions for Taiwan Futures Exchange

=head1 SYNOPSIS

  use Finance::TW::TAIFEX;

  my $taifex = Finance::TW::TAIFEX->new();

  $taifex->is_trading_day(); # is today a trading day?

  my $date = DateTime->now;
  $taifex->daily_futures_uri($date);
  $taifex->daily_options_uri($date);

  $taifex->contract('TX', '201001')->settlement_date;
  $taifex->product('TX')->near_term;
  $taifex->product('TX')->next_term;

=head1 DESCRIPTION

Finance::TW::TAIFEX provides useful helper functions for the Taiwan
Future Exchanges.

=head1 METHODS

=head2 product NAME

Returns the L<Finance::TW::TAIFEX::Product> object represented by NAME.

Currently supported product names:

=over

=item Futures

TX MTX TE TF T5F MSF CPF XIF GTF

=item Options

TXO TEO TFO MSO XIO GTO

=back

=head2 has_product NAME

Checks if the given product exists.

=cut

my $Strp = DateTime::Format::Strptime->new( pattern => '%F', time_zone => 'Asia/Taipei');

sub BUILDARGS {
    my $class = shift;
    return { @_ } unless $#_ == 0;
    return { } unless $_[0];
    return { context_date => ref $_[0] ? $_[0] : $Strp->parse_datetime($_[0]) }
}

=head2 contract NAME YEAR MONTH

Returns the L<Finance::TW::TAIFEX::Contract> of the given product expires on YEAR/MONTH.

=cut

sub contract {
    my ($self, $name, $year, $month) = @_;
    unless ($month) {
        my $spec = $year;
        ($year, $month) = $spec =~ m/^(\d{4})(\d{2})$/
            or die "$spec doesn't look like a contract month";
    }

    die "unknown product $name"
        unless $self->has_product($name);

    return Finance::TW::TAIFEX::Contract->new( product => $self->product($name),
                                               year => $year,
                                               month => $month );
}

sub _read_cal {
    my ($self, $file) = @_;
    open my $fh, '<', $file or die "$file: $!" ;
    return [map { chomp; $_ } <$fh>];
}

=head2 calendar_for YEAR

Returns the trading calendar for YEAR.

=cut

sub calendar_for {
    my ($self, $year) = @_;
    $year ||= $self->context_date->year;

    return $self->calendar->{$year}
        if $self->calendar->{$year};

    my $dist_dir = File::Spec->rel2abs("../../../share", File::Basename::dirname($INC{"Finance/TW/TAIFEX.pm"}));
    $dist_dir = try { dist_dir('Finance-TW-TAIFEX') || 'share' } unless -e $dist_dir;
    $self->calendar->{$year} = $self->_read_cal("$dist_dir/calendar/$year.txt");
}

=head2 is_trading_day [DATE]

Checks if the given DATE is a known trading day.  Default DATE is the date in the current context.

=cut

sub is_trading_day {
    my ($self, $date) = @_;
    $date ||= $self->context_date;

    $self->_nth_trading_day($self->calendar_for($date->year), $date->ymd) != -1;
}

=head2 next_trading_day [DATE]

Returns the next known trading day in string after the given DATE.

=cut

sub next_trading_day {
    my ($self, $date) = @_;
    $date ||= $self->context_date;

    my $cal = $self->calendar_for($date->year);
    my $d = $date->ymd;
    my $nth = firstidx { $_ gt $d } @{$cal};

    if ($nth < 0) {
        return $self->calendar_for($date->year + 1)->[0];
    }

    return $cal->[$nth];
}

=head2 previous_trading_day [DATE]

Returns the previous known trading day in string after the given DATE.

=cut

sub previous_trading_day {
    my ($self, $date, $offset) = @_;
    $date ||= $self->context_date;
    $offset ||= -1;

    my $cal = $self->calendar_for($date->year);
    my $nth = $self->_nth_trading_day($cal, $date->ymd);
    die "$date not a known trading day"
        if $nth < 0;

    if ($nth + $offset < 0) {
        return $self->calendar_for($date->year - 1)->[-1];
    }

    return $cal->[$nth + $offset];
}

sub _nth_trading_day {
    my ($self, $cal, $date) = @_;
    firstidx { $_ eq $date } @{$cal}
}

=head2 daily_futures_uri DATE

Returns the URI of the official TAIFEX futures trading records for DATE.

=cut

sub daily_futures_uri {
    my ($self, $date) = @_;
    $date ||= $self->context_date;
    return "http://www.taifex.com.tw/DailyDownload/Daily_@{[ $date->ymd('_') ]}.zip";
}

=head2 interday_futures_request($product, [$DATE])

Returns a HTTP::Request object that fetches futures monthly interday
csv file for $product of $DATE.

=cut

sub interday_futures_request {
    my ($self, $product, $date) = @_;
    $date ||= $self->context_date;
    my $from = $date->clone->truncate( to => 'month' );
    my $to = $from->clone->add( months => 1 )->subtract( days => 1 );
    return POST 'http://www.taifex.com.tw/chinese/3/3_1_2dl.asp',
      [ goday          => '',
        DATA_DATE      => $from->ymd('/'),
        DATA_DATE1     => $to->ymd('/'),
        DATA_DATE_Y    => $from->year,
        DATA_DATE_M    => $from->month,
        DATA_DATE_D    => $from->day,
        DATA_DATE_Y1   => $to->year,
        DATA_DATE_M1   => $to->month,
        DATA_DATE_D1   => $to->day,
        commodity_id2t => '',
        COMMODITY_ID   => $product ];
}

=head2 interday_options_request($product, [$DATE])

Returns a HTTP::Request object that fetches options monthly interday
csv file for $product of $DATE.

=cut

sub interday_options_request {
    my ($self, $product, $date) = @_;
    $date ||= $self->context_date;
    my $from = $date->clone->truncate( to => 'month' );
    my $to = $from->clone->add( months => 1 )->subtract( days => 1 );
    return POST 'http://www.taifex.com.tw/chinese/3/3_2_3_b.asp',
      [ goday          => '',
        DATA_DATE      => $from->ymd('/'),
        DATA_DATE1     => $to->ymd('/'),
        DATA_DATE_Y    => $from->year,
        DATA_DATE_M    => $from->month,
        DATA_DATE_D    => $from->day,
        DATA_DATE_Y1   => $to->year,
        DATA_DATE_M1   => $to->month,
        DATA_DATE_D1   => $to->day,
        COMMODITY_ID   => $product.'%' ];
}


=head2 daily_options_uri DATE

Returns the URI of the official TAIFEX options trading records for DATE.

=cut

sub daily_options_uri {
    my ($self, $date) = @_;
    $date ||= $self->context_date;
    return "http://www.taifex.com.tw/OptionsDailyDownload/OptionsDaily_@{[ $date->ymd('_') ]}.zip";
}

=head2 ensure_rpt DIR, TYPE, PREFIX

=cut

sub ensure_rpt {
    my ($self, $rpt_dir, $type, $prefix) = @_;
    my $date = $self->context_date;
    my $rpt_f = "$rpt_dir/".$date->ymd('-').".rpt";

    unless (-s $rpt_f) {
        my $rpt = $prefix.$date->ymd('_' ).".rpt";
        my $f = $self->can('daily_'.$type.'_uri') or die "unknown type: $type";
        my $url = $self->$f();
        my $tmp = "/tmp/taifex-$type-".$date->ymd('-').".zip";
        unless (-s $tmp) {
            my $rc = getstore($url => $tmp);

            die("failed to fetch $url: $rc")
                if HTTP::Status::is_error($rc);
        }
        my $zip = Archive::Zip->new();
        unless ( $zip->read( $tmp ) == AZ_OK ) {
            unlink( $tmp );
            die("Unable to read zip file $tmp");
        }
        unless ( $zip->extractMember($rpt, $rpt_f) == AZ_OK ) {
            die("Unable to extract $rpt from $tmp");
        }
    }
}

=head1 CAVEATS

The URI returned by C<daily_futures_uri> and C<daily_options_uri> are only valid for the last 30 trading days per the policy of TAIFEX.

=head1 AUTHOR

Chia-liang Kao E<lt>clkao@clkao.orgE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<http://www.taifex.com.tw/>

=cut

__PACKAGE__->meta->make_immutable;
no Any::Moose;
1;

